[pnnl/SHAD#22] Simple wrapper to allocate one instance per Locality.

mminutoli opened the pull request:
This pull request is a proposal to solve issue #22.  It will be used as a building block to implement reductions.

VitoCastellana commented on include/shad/data_structures/one_per_locality.h:
```diff
@@ -0,0 +1,81 @@
+
+  /// @brief Constructor.
+  template <typename... Args>
+  explicit OnePerLocality(ObjectID oid, Args... args)
```
This should be protected. Also, the documentation for the create method should be added.

mminutoli commented:
@VitoCastellana: I have updated the pull request to incorporate your comments.  I added a test storing and reading a different value on each locality.  It will help when we will go back to multi-locality systems.

VitoCastellana commented on include/shad/data_structures/one_per_locality.h:
```diff
@@ -0,0 +1,98 @@
+  /// @return A shared pointer to the local OnePerlocality instance.
+#ifdef DOXYGEN_IS_RUNNING
+  template <typename... Args>
+  static ShadHashmapPtr Create(Args... args);
```
:D Lovely cut & paste :-P

VitoCastellana commented on test/unit_tests/data_structures/one_per_locality_test.cc:
```diff
@@ -0,0 +1,93 @@
+        ASSERT_EQ(static_cast<int>(*localPtr), expectedValue);
+      },
+      anInt->GetGlobalID());
+}
```
anInt is not Destroyed properly

mminutoli replied:
nice catch!

mminutoli commented:
@VitoCastellana It should be good to be merged now.
